### ISSUE TITLE
ui-switch: state request

### ISSUE THREAD

**Issue by burner-** [FlowFuse/node-red-dashboard#1124]
### Description

allow request switch state with incoming message. Example with msg.get=true or something similar. 

Example usecase:
![image](https://github.com/user-attachments/assets/89796206-bf30-4d0a-94aa-aa185eb3cb35)
Currently this is possible by setting main power status with flow.set but that feature will save coding at many cases and make logic more readable at designer. 

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

**Comment by joepavitt** [comment]
Instead of the function node - could you not use a "Change" node which would improve the readability (and make it more low-code)

I'm not too keen on the idea of having a `msg.get` approach to influence the behaviour of the node - it's not a pattern I've seen in any other Node-RED nodes.

**Comment by burner-** [comment]
do you mean that use Change node to save switch status to flow and then at another place Change node to pick it up from flow store? Yes I can do it but example today I had case where I need to do 12 "off" switches what will force off some relays from automation. For that I need manually configure 12pcs flow get/set checks. It is quite fast just big unreadable mesh. With that feature I can just set those switched to flow and then push automation data trough those switches. Then match that if msg.automation === true && msg.payload === true return true;. It keep flow nice and clear to read when action goes through flow instead of storing states manually and then picking them up at another place.  By my understanfing continous flows are more how it is ment to use than short flows where events get stored to states and then read back to events at another place. Multiple scattered state<>event conversions make things always less robust and less readable.

**Comment by joepavitt** [comment]
I may need some clarity in your explanation of this piece:

> With that feature I can just set those switched to flow and then push automation data trough those switches. Then match that if msg.automation === true && msg.payload === true return true;.

But just to show a quick demo using a _single_ `change` node, alongside something like `msg.topic` to map the data:

<img width="458" alt="Screenshot 2024-07-26 at 18 19 39" src="https://github.com/user-attachments/assets/e7dd0f73-5706-4172-a15b-2b8936a2aeb8">

<img width="363" alt="Screenshot 2024-07-26 at 18 19 48" src="https://github.com/user-attachments/assets/211b8e23-5762-4eb5-8bbc-07e3685bfa55">

>  By my understanding continuous flows are more how it is meant to use than short flows where events get stored to states and then read back to events at another place. Multiple scattered state<>event conversions make things always less robust and less readable.

They are certainly different patterns for sure, but neither are _wrong_. They're both valid ways of using Node-RED

**Comment by joepavitt** [comment]
Now just reading back, I'm realising the use case, you have two switches, and when you're triggering the first switch, you want to also _read_ the state of the second switch, and combine those into a single model?

**Comment by joepavitt** [comment]
We _could_ require even less than a `msg.get`? What about if you pass in _any_ payload that isn't a valid on/off state, then we just read the value and return that?

**Comment by colinl** [comment]
If this is done for the switch node then exactly the same argument could be made for any input node. The node red way is to use Join nodes to retain the current state and merge it with any events as necessary.

**Comment by burner-** [comment]
> If this is done for the switch node then exactly the same argument could be made for any input node. The node red way is to use Join nodes to retain the current state and merge it with any events as necessary.

You are right that this same argument is valid for all state based input nodes. And I think it just make this feature more relevant. Event based input nodes wont apply. Generally speaking it is usually good to have one authorative place to store state. And because flow store cannot send event when it get changed, ui node is quite good place for it. At least it already implement event sending and showing current state to user. 


> We _could_ require even less than a `msg.get`? What about if you pass in _any_ payload that isn't a valid on/off state, then we just read the value and return that?

That may work also. I just wonder that will it cause situation where result can be un excepted? But I agree that it may have good or even better approach. Then user can just send message without payload and switch will set its current payload to reply. That will allow following use:
![image](https://github.com/user-attachments/assets/faddda84-d53c-40ff-9283-de70bc26e313)

**Comment by burner-** [comment]
fyi, one more usecase to bring up that this would be good for all input nodes. 
![image](https://github.com/user-attachments/assets/b6fedbf4-6f5c-4f2f-b219-5547699552c0)
Basically that allow augment on/off state with brighness without saving it first to flow.

**Comment by colinl** [comment]
If it is going to MQTT then use a Retained topic and MQTT will remember it for you.  Then you can feed the value in from an MQTT node on startup to initialise the switch to its saved state.

**Comment by burner-** [comment]
> If it is going to MQTT then use a Retained topic and MQTT will remember it for you. Then you can feed the value in from an MQTT node on startup to initialise the switch to its saved state.

I think you missed point quite thoroughly. And it is not that I cannot do it with lot of different ways. Surely not with retained message but anyway its not point of whole thing. It was just idea how to make dashboard components for more zero code friendly but maybe it was just mistake to even try contribute.

**Comment by joepavitt** [comment]
> Basically that allow augment on/off state with brighness without saving it first to flow.

I'm not sure I understand, what would you want the payload going into the MQTT node to be here?

**Comment by colinl** [comment]
I was responding to this comment, not your general suggestion.

> that allow augment on/off state with brighness without saving it first to flow.

There is no need to save it to the flow at all, when the switch is clicked it will be sent to MQTT and saved there.  No need for flow context at all.

**Comment by burner-** [comment]
> I was responding to this comment, not your general suggestion.
> 
> > that allow augment on/off state with brighness without saving it first to flow.
> 
> There is no need to save it to the flow at all, when the switch is clicked it will be sent to MQTT and saved there. No need for flow context at all.

How you plan to augment on/off event with brighness data in mqtt without taking it to flow context first? At that example brigness was augmented to message so that it is possible to format that to one message what contains brighness and on/off information. That is quite common usecase. I think you still miss whole point some how. Point is not get one status retained. Point is do flows where multiple inputs are glued together to one output or multiple inputs affects with each other as logic manner.

**Comment by burner-** [comment]
> > Basically that allow augment on/off state with brighness without saving it first to flow.
> 
> I'm not sure I understand, what would you want the payload going into the MQTT node to be here?

it is just putput example. It can be even debug node but mostly everything end to devices via mqtt so it was simple example flow. Yes maybe it need first change node or template node to format outputs suitable for target device.
